src/IECoreScene/MeshAlgoSplit.cpp

```diff
@@ -0,0 +1,545 @@
+// primitive variables.
+//
+// It is implemented as a vector of fixed size blocks of memory spanning the entire range of
+// original ids.  This is wasteful of memory when the number of indices is very low relative
+// to the range of original ids ( ie. you are splitting an extremely large mesh into extremely
+// small pieces ), but it much more efficient to just index into a location than it is to
+// hash an integer to use it as a hashmap key.
+//
+// It would be possible to implement a much more compact version storing only 1 bit per id,
+// with separate counts every 64 or 128 ids, which would perform better on very large
+// meshes being split into very small pieces, but that doesn't help much in average cases,
+// and is a fair bit more complicated.  If we encounter issues with performance when
+// splitting into tiny meshes, the simplest solution is probably to switch to an unordered_map
+// when numIndices is much smaller than numOriginalIds - the break even point for performance
+// seems to be when numOriginalIds is about 10 000 times greater than numIndices - or actually
+// much higher when the indices are fairly coherent ( ie. the ids which are selected fall
+// mainly in the same range, so many are in the same block ), which is common for most ways of
+// producing meshes.
+
+class Reindexer
+{
+public:
+
+	// Construct a Reindexer
+	//
+	// numOriginalIds : determines the highest integer that may appear in the indices
+	// numIndices : how many indices will be added.  You must call addIndex() this many times
+	// blockSize : a performance tuning value determining how large the blocks that are
+	//     allocated to hold ids are.  Should be left at default.
+	//
+	Reindexer( int numOriginalIds, int numIndices, int blockSize = 1024 ) :
+		m_newIndicesData( new IntVectorData() ),
+		m_newIndices( m_newIndicesData->writable() ),
+		m_blockSize( blockSize ),
+		m_fromOldIds( ( numOriginalIds - 1 ) / blockSize + 1 ),
+		m_numIdsUsed( 0 )
+	{
+		m_newIndices.reserve( numIndices );
+	}
+
+	// Add an index - if the indexed id is not yet part of the output ids, it will be included
+	void addIndex( int id )
+	{
+		// Determine which block to use, and the index within that block
+		int blockId = id / m_blockSize;
+		int subIndex = id % m_blockSize;
+
+		auto &block = m_fromOldIds[ blockId ];
+
+		if( block )
+		{
+			// Block for this id is already allocated
+
+			int &r = (*block)[ subIndex ];
+			if( r == -1 )
+			{
+				// Id isn't used yet, we need to set this location in the block, and use it
+				r = m_numIdsUsed;
+				m_numIdsUsed++;
+			}
+
+			m_newIndices.push_back( r );
+		}
+		else
+		{
+			// Block isn't allocated.  Allocate the block, and set the entry for this id
+
+			block = std::make_unique< std::vector<int> >( m_blockSize, -1 );
+			(*block)[subIndex] = m_numIdsUsed;
+			m_newIndices.push_back( m_numIdsUsed );
+			m_numIdsUsed++;
+		}
```

Comment:
Is the duplicate code between these two branches definitely worth it? It seems we could just allocate the block if necessary and then use the same code to update indices? And the only extra cost would be single comparison to `-1` each time a block is allocated?